ministryofjustice/analytics-platform-control-panel
# Create table to hold allowed IP ranges

## :memo: Summary
This PR completes issue #ANPL-1036.


This PR adds a new table to hold the allowed IP ranges. These will be used to generate allowed IP ranges for individual apps in follow-on work (see ANPL-1034). This table is structured as specified in ANPL-1035/[Miro board](https://miro.com/app/board/uXjVPcxSzik=/).

Note that this PR also includes a set of migrations that are needed following a version upgrade of Django several months ago. This sets default types for IDs in several tables, and also changes the default `Tool` type to EKS-compatible rather than kOps-compatible.


The changes in this PR are needed because we are refactoring our process of creating an allowlist of IP addresses for each app. This PR creates the table where the allowlist will be stored.


Merging this PR will have the following side-effects:
- None expected.

## :mag: What should the reviewer concentrate on?
- Check migration `0017` for any unintended side-effects.
- What do you think of the [`django-auditlog`](https://django-auditlog.readthedocs.io/en/latest/index.html) package? I introduced it as the idea to track changes was of interest when I mentioned it on the Miro board (linked above). Note that as history can be tracked via `django-auditlog`, the `IPAllowlist` model does not include `created` and `modified` columns (which are present in most of the other models as they are based on `TimeStampedModel`).
  - Following @ymao2's feedback, I have switched to [`django-simple-history`](https://django-simple-history.readthedocs.io/en/latest/index.html). This offers similar functionality but the history is stored in a separate table to the `IPAllowlist` model. The testing guide below has been updated to reflect the change of package. I have also added `created` and `modified` via `TimeStampedModel` back in.

## :technologist: How should the reviewer test these changes?
- Set up your local Control Panel following the guide in `doc/running.md`.
- Update your installed packages with `pip install -r requirements.txt` (this will add ~`django-auditlog`~ `django-simple-history`).
- Make sure you have an instance of `User` and `Tool` in your database
- Run migrations with `python manage.py migrate`
- Check that migration `0017` hasn't done anything unexpected to your tables (check `Tool`, `User` and any others you have locally).
- Try creating a new `IPAllowlist`, make some changes and look at the history. Something like this should get you started, and after that make some changes and see how they get recorded:
```python
from controlpanel.api.models import IPAllowlist

ipa = IPAllowlist(name="Test", allowed_ip_ranges="123")    # will add validator for IP range format in ANPL-1037

IPAllowlist.history.all()                   # should be empty
ipa.save()
IPAllowlist.history.all()                   # one entry
ipa.contact = "a name"
IPAllowlist.history.all()                   # still one entry
ipa.save()
IPAllowlist.history.all()                   # updated with second entry

diff = IPAllowlist.history.first().diff_against(IPAllowlist.history.last())
for change in diff.changes:
    print("{}: \'{}\' -> \'{}\'".format(change.field, change.old, change.new))
```

## :books: Documentation status

- [x] Documentation will be added in the future because we can add this all in one go at the end of ANPL-1034.
